### ISSUE TITLE
make a plugin system

### ISSUE THREAD

**Issue by ajm188** [ajm188/slack#32]
Here is what I'm thinking.

If we use the following type definitions:

``` go
// library code: plugin.go
package slack

type Plugin interface {
    Name() string
    CanLoad() bool
    Load(*Bot)
}
type PluginFunc func() Plugin
```

Then we can implement a plugin like so:

``` go
// library code: plugins/github/github.go
package github

import (
    "fmt"
    "os"
)

type OpenIssuePlugin struct{}
func OpenIssue() *OpenIssuePlugin { return &OpenIssuePlugin{} }

func (gh *OpenIssuePlugin) Name() string {
     return "GH: OpenIssue"
}

func (gh *OpenIssuePlugin) CanLoad() (ok bool) {
    ok = true
    envVars := []string{"GH_CLIENT_ID", "GH_CLIENT_SECRET", "GH_ACCESS_TOKEN"}
    for _, eVar := range envVars {
        if os.Getenv(eVar) == "" {
            fmt.Println("Missing env var %s.", eVar)
            ok = false
        }
    }
    return
}

func (gh *OpenIssuePlugin) Load(bot *slack.Bot) {
     // do stuff to register handler
}
```

And a bot can consume this code as follows:

``` go
// main.go
package main

import (
    "github.com/ajm188/slack"
    "github.com/ajm188/slack/plugins/github"
)

func main() {
    bot := slack.NewBot("TOKEN")
    bot.UsePlugin(github.OpenIssue)
}
```

**Comment by ajm188** [comment]
If this works well, we can look into using [reflection](https://golang.org/pkg/reflect/) to load plugins from a config file.

**Comment by ajm188** [comment]
The library code builds. Maybe instead the `UsePlugin` function that I made up should look like:

``` go
func (b *Bot) UsePlugin(f PluginFunc, args ...interface{})
```

Then `f` will be stored to be called later. Calling `f` will actually "load" the plugin, and `f` will be called with `args`.

**Comment by ajm188** [comment]
I have a working compile of this, with some notable changes to what I've described above.

I need to clean up the mess I've made, write some tests, and then will submit a PR.

**Comment by brenns10** [comment]
I'm interested in helping get the plugin system off the ground, so maybe I'll assist on this.  But right now, my main thought that I wanted to contribute is the plugin interface ought to include some sort of help text feature so that you can ask slacksoc for help.

**Comment by ajm188** [comment]
Cool! Glad you're interested in helping out :)

Let's use this issue to hash out details of how this should work. And - as always - pull requests welcome :)

**Comment by ajm188** [comment]
@brenns10 I was thinking some more about this today.

I think we want something like

``` golang
type Plugin interface {
    Register(Bot) error
}
```

Then the plugin does whatever it needs to do to register itself with the bot. It should return an error if it runs into any issues (missing an environment variable, etc).

The bot should load up all the plugins that have registered before the bot starts up and connects to slack. I'm also thinking about a system where we have plugin loading options (i.e. my plugin is thread-safe, so go ahead and load it in a goroutine for more speed; or we don't need to crash the whole bot if this plugin fails to load, or something to that effect), but I'm not sure what that would look like yet. Maybe something like:

``` golang
type PluginRegistrationConfig struct {
    ThreadSafe bool
    Critical bool // this is a really bad name for the second thing I was talking about above
}

// Then something like
type Plugin interface {
    Register(Bot, *PluginRegistrationConfig)
}
myPlugin.Register(bot, &PluginRegistrationConfig{true, false})
```

Thoughts?

**Comment by brenns10** [comment]
Regarding "Thread Safe" config, how important do you think startup time
will be on the bot, and do you anticipate it will be a problem? Sounds like
premature optimization to consider loading plugins asynchronously unless
you've observed it to be a performance issue. Regarding "Critical" config,
this one makes sense, but I think it would be good to think through error
management. For instance, if we're allowing non-critical plugins to error
out without crashing the entire application, we had better expect that
these plugins can clean up their handlers and everything, probably with
some sort of Unregister() function in the plugin interface.

**Comment by ajm188** [comment]
> how important do you think startup time will be on the bot, and do you anticipate it will be a problem?

Not really. In an ideal world, this bot library becomes wildly popular, and users want to load a zillion plugins.

> Regarding "Critical" config, this one makes sense, but I think it would be good to think through error management.

I agree. My thoughts are as follows:
- Setting "Critical" is up to the user of the plugin, not the author.
  - This means that it is a measure of the user's level of trust in the plugin. If the user knows that the plugin cleans up nicely after itself, and doesn't care _too_ much either way whether the plugin is loaded or not, then they can set `Critical = false`.
  - If they don't trust the plugin, or the plugin has known issues with dirtying up the bot if it fails halfway through registering, then they can set `Critical = true` and avoid the whole issue.

Maybe it makes sense for Critical to always be `true` in which case this is a moot point. I have no idea this early on.
